<a href="https://colab.research.google.com/github/guscldns/TestProject/blob/main/0612/0612_02_1_RNN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [4]:
import os 
import numpy as np
import matplotlib.pyplot as plt

In [5]:
import tensorflow as tf
tf.__version__

'2.12.0'

# Keras RNN 이해하기

- https://www.tensorflow.org/api_docs/python/tf/keras/layers/SimpleRNN

In [6]:
from tensorflow.keras.layers import SimpleRNN 

In [ ]:
SimpleRNN(10)

In [ ]:
inputs = np.random.random([32, 10, 8]).astype(np.float32)
# input = 3D 텐서, 모양 . [batch, timesteps, feature]
# 32 데이터 개수(batch), 10은 문장 내 단어 개수(timesteps), 8 (feature, 텍스트에서 단어 벡터의 크기)
# 문장 내 단어가 없는 경우 빈 행렬로 값을 채워준다

simple_rnn = tf.keras.layers.SimpleRNN(4)# ht를 return 
output = simple_rnn(inputs) 
# output : 마지막 시점의 hidden vector를 출력 (32,4) >> 10*8 벡터가 4로 바뀜

In [ ]:
print(inputs[0].shape) # 0시점 x0 , x1, x2, ..., x9 (크기가 8짜리 벡터)
# >> RNN
print(output[0].shape)
# 10,8 이 들어가서 4, 의 벡터로 나온다

In [ ]:
inputs = np.random.random([32, 10, 8]).astype(np.float32)
simple_rnn = tf.keras.layers.SimpleRNN(4, return_sequences=True)
# many to many 등 다중으로 나와야 할 경우 시점 정보(h0...ht)가 살아있어야 한다 = return_sequences=True로 살려야 한다
output = simple_rnn(inputs)

In [ ]:
output

In [ ]:
# X : John loves Jane 
# y : S V O 


# Ht -> S, Ht -> V, Ht -> O 
# 시점정보를 주면 H0 -> S, H1 -> V, Ht -> O (이 편이 학습을 더 잘한다)

In [7]:
inputs = np.random.random([32, 10, 8]).astype(np.float32)
simple_rnn = tf.keras.layers.SimpleRNN(4, return_sequences=True, return_state = True) # 둘 다 True로 하면 전체 시점의 히든 벡터가 출력됨
output, ht = simple_rnn(inputs)

In [8]:
output # shape=(32, 10, 4)

<tf.Tensor: shape=(32, 10, 4), dtype=float32, numpy=
array([[[-0.6630791 , -0.3428777 , -0.45282987,  0.11329813],
        [-0.49131468,  0.5947519 , -0.74654776,  0.02015117],
        [-0.8461054 , -0.07381713, -0.71039605, -0.5072021 ],
        ...,
        [-0.27041733,  0.3385326 , -0.6478172 ,  0.05191475],
        [-0.5430749 , -0.17305389, -0.3365632 , -0.5206878 ],
        [-0.864594  ,  0.2998157 , -0.9638729 ,  0.62137276]],

       [[-0.73711675, -0.28694242, -0.38040775,  0.04889037],
        [-0.6633584 ,  0.3070362 ,  0.05222382,  0.38923743],
        [-0.5831184 ,  0.0677215 , -0.59732807, -0.5957278 ],
        ...,
        [-0.8537511 ,  0.47793534, -0.91424346,  0.41936567],
        [-0.9257834 ,  0.7229155 , -0.8577859 , -0.2896206 ],
        [-0.91125965,  0.63435626, -0.33909804, -0.1157968 ]],

       [[-0.86861134, -0.5224898 , -0.533459  ,  0.446595  ],
        [-0.1667595 ,  0.6785427 , -0.4397136 , -0.04556594],
        [-0.9413238 , -0.29844064, -0.8509589 , -

In [9]:
ht# shape=(32, 4)

<tf.Tensor: shape=(32, 4), dtype=float32, numpy=
array([[-0.864594  ,  0.2998157 , -0.9638729 ,  0.62137276],
       [-0.91125965,  0.63435626, -0.33909804, -0.1157968 ],
       [-0.907511  ,  0.35511798, -0.69433844,  0.00363444],
       [-0.777663  ,  0.5805959 , -0.97616   , -0.35445985],
       [-0.71356994,  0.5528299 , -0.9147907 , -0.2822579 ],
       [-0.8831751 ,  0.5371211 , -0.40997466,  0.48866612],
       [-0.97554827,  0.00241589, -0.07607556, -0.24485454],
       [-0.9454134 ,  0.6186545 , -0.7582003 , -0.39372897],
       [-0.8741206 ,  0.8125893 , -0.5852022 , -0.41528004],
       [-0.68299943,  0.7065444 , -0.92657167, -0.73945653],
       [-0.8033803 ,  0.8391903 , -0.92645705, -0.5016358 ],
       [-0.55548257,  0.8444395 , -0.9061835 , -0.73266774],
       [-0.78166866,  0.2155573 , -0.9561071 ,  0.5001253 ],
       [-0.73470914,  0.78349614, -0.77031374, -0.7016787 ],
       [-0.92457587,  0.7831085 , -0.95929   , -0.38465342],
       [-0.7744559 ,  0.8635837 , -0

In [ ]:
output[0] # 맨마지막 값이랑 ht[0]이랑 값이 똑같다

<tf.Tensor: shape=(10, 4), dtype=float32, numpy=
array([[-0.2369311 , -0.21232551,  0.5750342 ,  0.92289734],
       [-0.6524617 , -0.452945  ,  0.96605444,  0.9834321 ],
       [-0.6171921 , -0.33961102,  0.9380261 ,  0.9978312 ],
       [-0.78218406, -0.49791923,  0.97936624,  0.99841714],
       [-0.59237236, -0.5842285 ,  0.8234757 ,  0.9896057 ],
       [-0.57326937, -0.23393983,  0.8950576 ,  0.9959948 ],
       [-0.57525194, -0.47060463,  0.94015807,  0.99144804],
       [-0.5078638 , -0.53244436,  0.9699914 ,  0.99484104],
       [-0.5709194 , -0.24397987,  0.9389674 ,  0.9969233 ],
       [-0.562725  , -0.68468213,  0.96814805,  0.99872094]],
      dtype=float32)>

In [ ]:
ht[0]

<tf.Tensor: shape=(4,), dtype=float32, numpy=array([-0.562725  , -0.68468213,  0.96814805,  0.99872094], dtype=float32)>

In [ ]:
simple_rnn = tf.keras.layers.SimpleRNN(4, return_sequences=True, return_state=True)

whole_sequence_output, final_state = simple_rnn(inputs)

RNN의 input, output, weight을 이해해봅시다.

In [11]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, GRU, LSTM, Dense

### 문제1

In [ ]:
model = Sequential()
model.add(SimpleRNN(5, input_shape=(2,10)))
# input_shape=(2,10) 앞에 timestep, 뒤가 단어벡터 크기 
# model.add(SimpleRNN(3, input_length=2, input_dim=10))와 동일
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 simple_rnn (SimpleRNN)      (None, 5)                 80        
                                                                 
Total params: 80
Trainable params: 80
Non-trainable params: 0
_________________________________________________________________


In [ ]:
# ht = tang (Wx* xt + Wh *h(t-1) + b)
# ht == SimpleRNN(5 << = 5*1
# 식에 대입하려면 5*1(=Wx* xt) + 5*1(=Wh *h(t-1) + 5*1(=b) 로 나타나야 함
# xt = input_shape=(2,10)의 10 == 10*1

# Wx(5*10), Wh(5*5), b(5*1) 다 합하면 80 

모델 summary를 보고 생각해봅시다. 
- input은 어떤 모양인가요? 
 : 무조건 3D, 데이터 개수, timestep, 단어벡터 크기 ( == input dimension)
- output은 어떤 모양인가요? 
 : 2D, 데이터 개수, 히든 벡터 크기 ( == output dimension)

In [ ]:
# input : 3D -> 데이터 개수, timestep, 단어벡터 크기(==input dimension)
# output : 2D -> 데이터 개수, 히든 벡터의 크기(==output dimension)

### 문제2

In [ ]:
model = Sequential()
# The output of RNN will be a 3D tensor of shape (batch_size, input_dim)
model.add(SimpleRNN(5, batch_input_shape=(8,2,10)))
# 벡터 size 는 input, output dimension이 결정한다, 각각 결정 X 둘이 결정
model.summary()
# input, output dimension 모두 동일 >> paramiter 개수 동일

### 문제3

- 중간 hidden state를 모두 리턴하는 return_sequences를 True로 설정합니다. 

In [ ]:
model = Sequential()

# The output of RNN will be a 3D tensor of shape (batch_size, timesteps, input_dim)
model.add(SimpleRNN(5, input_dim=10, return_sequences=True))
model.summary()
# input, output dimension 모두 동일 >> paramiter 개수 동일
# 3차원으로 나타날 지언정 param 개수는 동일하다

무엇이 달라졌나요? 

In [ ]:
# 모델 복잡도는 input dimension과 output dimiension이 결정한다

이번엔 파라미터의 개수를 세어보세요. 어떻게 계산하나요? 

- Wxh = 
- Whh = 
- bias(hidden)  =

### 문제4

In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, Dense

model = Sequential()
model.add(SimpleRNN(7, input_shape=(100,5), return_sequences = True)) # 91
# ht = tang (Wx* xt + Wh *h(t-1) + b) # xt는 input_shape=(100,5) 뒤의 5 = 5*1
# 7*1 = (7*5*5*1) + (7*7 *7*1) + (7*1) >>  Wx = 7*5(35)+ Wh = 7*7(49)+ b = 7*1(7) = 91 (35+49+7)
model.add(SimpleRNN(3)) # 33
# ht = tang (Wx* xt + Wh *h(t-1) + b) << 여기 xt는 위의 층의 ht임
# 3*1 (3*7*7*1) + (3*3 *3*1) + (3*1) >>  Wx = 3*7(21)+ Wh = 3*3(9)+ b = 3*1(3) = 33 (35+49+7)
model.add(Dense(2)) # 12
# y = softmax (Wy * x +b) >> X는 위의 ht = 3*1
# y = 3*3 +3*1 + 3*1 >> Wy = 3*3(9)+ b = 3*1(3) = 12
model.summary()
# 알아서 자동으로 흘러가는 구조가 시퀀설 구조

위의 summary를 보고 직접 계산해보세요

Wx =  1층 : 35, 2층 21, 3층 9

Wh =  1층 : 49, 2층 9, 3층 없음

bh =  1층 : 7, 2층 3, 3층 3


In [ ]:
## 오류가 날 때 차원으로 오류가 나기 때문에 어떤 차원이 input인지 알아둬야 한다

### 문제5

output layer를 추가하겠습니다. 

In [ ]:
model = Sequential()
model.add(SimpleRNN(5, input_shape = (2, 10)))
model.add(Dense(1))
model.summary()

이번에는 파라미터 개수가 몇 개인가요? 

In [ ]:
Wyht + b = y 

# Keras로 RNN 구현하기

- Keras RNN : https://keras.io/api/layers/recurrent_layers/simple_rnn/

### 1. 단층 Simple RNN

우선 RNN과 LSTM을 테스트하기 위한 임의의 입력을 만듭니다.

In [28]:
# 단어 벡터
John = [1,0,0,0]
loves = [0,1,0,0]
cats = [0,0,1,0]
dogs = [0,0,0,1]

In [29]:
# X 데이터

train_X = np.array([
    [ John, loves, cats ],
    [ John, loves, dogs ]
]).astype(np.float32)

In [30]:
# 타겟 벡터 

S = [0] # subject
V = [1] # verb
O = [2] # object

idx2tag = ['S', 'V', 'O']

In [31]:
# Y 데이터
train_Y = np.array([[S, V, O], [S, V, O]]).astype(np.float32)

print("train_y", train_Y)
print("train_X의 shape", train_X.shape) 
# train_X의 shape (2, 3, 4) >> 
# 2차원([[John, loves, cats],[John, loves, dogs]]), 데이터 3개([ John, loves, cats ]) 들어있고 벡터 크기 4(John = [1,0,0,0])
print("train_Y의 shape", train_Y.shape)
# train_Y의 shape (2, 3, 1)
# 2차원 : ([[S, V, O], [S, V, O]]), 데이터 3개, 벡터 크기 1(S = [0])

train_y [[[0.]
  [1.]
  [2.]]

 [[0.]
  [1.]
  [2.]]]
train_X의 shape (2, 3, 4)
train_Y의 shape (2, 3, 1)


In [ ]:
train_X

In [ ]:
train_X.shape

(2, 3, 4)

In [ ]:
train_Y

array([[[0.],
        [1.],
        [2.]],

       [[0.],
        [1.],
        [2.]]], dtype=float32)

- 단어 벡터의 차원은? 
- 문장의 길이는? 
- output 클래스 개수는? 

위 입력은 단어 벡터의 차원은 4이고, 문장의 길이가 3인 경우를 가정한 입력입니다. 
다시 말해 3번의 시점(timesteps)이 존재하고, 각 시점마다 4차원의 단어 벡터가 입력으로 사용됩니다. 
     
그런데 RNN은 2D 텐서가 아니라 3D 텐서를 입력을 받습니다 -> (batchsize, timesteps, features)  = (batchsize, timesteps, input_dim)  
즉, 배치크기 2로 설정하여 입력합니다. 

In [27]:
# 하이퍼 파라미터
num_units = 3
input_dim = 4 # 단어 벡터 크기
sequence_length = 3 # 한 문장에 단어 개수 
learning_rate = 0.1


# 기본 RNN 
rnn = SimpleRNN(num_units, input_shape = (sequence_length, input_dim), return_sequences = True)
# h0, h1, h2 를 다 알아야 해서 return_sequences = True를 해줌 (many to many 문제라서)
# rnn = SimpleRNN(5, return_sequences=False, return_state=False)와 동일.
hidden_state = rnn(train_X)

print('hidden state : {}, shape: {}'.format(hidden_state, hidden_state.shape))

hidden state : [[[ 0.650981   -0.4265175   0.640633  ]
  [-0.5788596  -0.49438623  0.66779244]
  [-0.8622027  -0.43359882 -0.12346305]]

 [[ 0.650981   -0.4265175   0.640633  ]
  [-0.5788596  -0.49438623  0.66779244]
  [-0.9309214   0.04430077 -0.8464287 ]]], shape: (2, 3, 3)


(2, 3) 크기의 텐서가 출력되는데, 이는 마지막 시점의 은닉 상태입니다.     
기본적으로 return_sequences가 False인 경우에는 SimpleRNN은 마지막 시점의 은닉 상태만 출력합니다. 이번에는 return_sequences를 True로 지정하여 모든 시점의 은닉 상태를 출력해봅시다.

In [18]:
# return_sequences=True
rnn = SimpleRNN(num_units, return_sequences=True)
hidden_states = rnn(train_X)

print('hidden states : {}, shape: {}'.format(hidden_states, hidden_states.shape))

hidden states : [[[-0.41307232  0.6271849  -0.14366776]
  [-0.00372664 -0.44388834  0.14864713]
  [-0.83675456  0.03727749 -0.5249121 ]]

 [[-0.41307232  0.6271849  -0.14366776]
  [-0.00372664 -0.44388834  0.14864713]
  [-0.74678665 -0.72261196 -0.19151336]]], shape: (2, 3, 3)


return_sequences가 True일 경우 모든 시점에 대해서 은닉 상태의 값을 출력합니다.



In [19]:
# return_sequences=True && return_state=True
rnn = SimpleRNN(num_units, return_sequences=True, return_state=True)
hidden_states, last_state = rnn(train_X)

print('hidden states : {}, shape: {}'.format(hidden_states, hidden_states.shape))
print('last state : {}, shape: {}'.format(last_state, last_state.shape))

hidden states : [[[-0.29648376  0.6081849  -0.5885171 ]
  [ 0.01980543 -0.14215292  0.6032151 ]
  [ 0.6094724  -0.5715136  -0.7255894 ]]

 [[-0.29648376  0.6081849  -0.5885171 ]
  [ 0.01980543 -0.14215292  0.6032151 ]
  [-0.3125006  -0.86412144 -0.65956163]]], shape: (2, 3, 3)
last state : [[ 0.6094724  -0.5715136  -0.7255894 ]
 [-0.3125006  -0.86412144 -0.65956163]], shape: (2, 3)


첫번째 출력은 return_sequences=True로 인한 출력으로 모든 시점의 은닉 상태입니다. 두번째 출력은 return_state=True로 인한 출력으로 마지막 시점의 은닉 상태입니다

In [32]:
from tensorflow.keras import layers, models

model = models.Sequential() #모델 호출
model.add(layers.SimpleRNN(units= 5, # output 벡터의 크기 << 이 모델에서 현재 데이터로 units=만 사람이 바꿀 수 있다
                           input_shape = (3, 4), # time step, input 벡터의 크기, x 데이터에 맞춰야 함(timestep : 배열(시퀀스)의 길이)
                           return_sequences = True, # 모든 시점의 hidden state를 받아오기 , RNN 층 개수, 다중분류냐 이진분류나 등에 따라 사용
                           name='SimpleRNN-1')) # 해당 레이어 이름 지정


model.add(layers.Dense(units=3, # 3가지로 분류해야해서 3
                       activation='softmax', 
                       name='hidden-to-output')) # 다중 분류 출력층, 카테고리 3개 

model.compile(
    loss='sparse_categorical_crossentropy', 
    # y값이 라벨이라 sparse 붙어야 함
    optimizer='adam', 
    metrics = 'accuracy')

In [33]:
model.fit(train_X, train_Y, epochs= 50)

Epoch 1/50
1/1 [==============================] - 1s 1s/step - loss: 1.0348 - accuracy: 0.3333
Epoch 2/50
1/1 [==============================] - 0s 12ms/step - loss: 1.0292 - accuracy: 0.3333
Epoch 3/50
1/1 [==============================] - 0s 12ms/step - loss: 1.0235 - accuracy: 0.3333
Epoch 4/50
1/1 [==============================] - 0s 15ms/step - loss: 1.0180 - accuracy: 0.3333
Epoch 5/50
1/1 [==============================] - 0s 12ms/step - loss: 1.0124 - accuracy: 0.3333
Epoch 6/50
1/1 [==============================] - 0s 13ms/step - loss: 1.0069 - accuracy: 0.3333
Epoch 7/50
1/1 [==============================] - 0s 12ms/step - loss: 1.0014 - accuracy: 0.3333
Epoch 8/50
1/1 [==============================] - 0s 11ms/step - loss: 0.9959 - accuracy: 0.5000
Epoch 9/50
1/1 [==============================] - 0s 15ms/step - loss: 0.9905 - accuracy: 0.5000
Epoch 10/50
1/1 [==============================] - 0s 14ms/step - loss: 0.9851 - accuracy: 0.5000
Epoch 11/50
1/1 [==============

In [ ]:
# "John" -> model -> ????


X = np.array([[John, John, John]])
X.shape

(1, 3, 4)

In [ ]:
model.predict(X)

1/1 [==============================] - 0s 181ms/step


array([[[0.38968852, 0.25487605, 0.35543543],
        [0.4606964 , 0.2343276 , 0.30497602],
        [0.3839034 , 0.16858758, 0.447509  ]]], dtype=float32)

In [37]:
model.summary()

Model: "sequential_4"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 SimpleRNN-1 (SimpleRNN)     (None, 3, 5)              50        
                                                                 
 hidden-to-output (Dense)    (None, 3, 3)              18        
                                                                 
Total params: 68
Trainable params: 68
Non-trainable params: 0
_________________________________________________________________


In [38]:
model.weights # 모델의 weight를 알고 싶을 때 사용하는 함수

[<tf.Variable 'SimpleRNN-1/simple_rnn_cell_10/kernel:0' shape=(4, 5) dtype=float32, numpy=
 array([[ 0.06959531, -0.52854717, -0.14654705,  0.27457142,  0.48504537],
        [-0.6317964 , -0.5980677 ,  0.4007017 , -0.29166332, -0.5397489 ],
        [ 0.14714886, -0.7389745 ,  0.2724483 ,  0.6763485 , -0.41975278],
        [-0.15534993,  0.67234683, -0.4914142 ,  0.45404738,  0.21804516]],
       dtype=float32)>,
 <tf.Variable 'SimpleRNN-1/simple_rnn_cell_10/recurrent_kernel:0' shape=(5, 5) dtype=float32, numpy=
 array([[-0.25555283,  0.58974123,  0.6040706 , -0.09966107,  0.2708081 ],
        [-0.8654383 , -0.3620915 , -0.1666469 ,  0.17646585,  0.4102106 ],
        [-0.18297748,  0.13068764,  0.20720346,  0.91429895, -0.39749348],
        [-0.09170093,  0.59679455, -0.7876978 ,  0.06110502,  0.05710199],
        [ 0.5054425 , -0.15574482, -0.05588553,  0.42416498,  0.69703937]],
       dtype=float32)>,
 <tf.Variable 'SimpleRNN-1/simple_rnn_cell_10/bias:0' shape=(5,) dtype=float32, num

In [ ]:
predictions = model.predict(train_X)


for i, prediction in enumerate(predictions):
    print(prediction)
    print(np.argmax(prediction, axis=1))

    result_str = [idx2tag[c] for c in np.argmax(prediction, axis=1) ]
    
    print("\tPrediction str: ", "".join(result_str), end = '\n\n\n')


[[0.7663143  0.107223   0.12646265]
 [0.14568445 0.714073   0.14024259]
 [0.09099198 0.18833458 0.7206734 ]]
[0 1 2]
	Prediction str:  SVO


[[0.5298186  0.2547431  0.21543826]
 [0.11390428 0.6944526  0.19164307]
 [0.09559339 0.17389421 0.7305124 ]]
[0 1 2]
	Prediction str:  SVO




모델을 이해하려면 weight 개수(W0\~Wn)를 세어봐야 합니다.

In [ ]:
model.summary()

Model: "sequential_10"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 SimpleRNN-1 (SimpleRNN)     (None, 3, 3)              24        
                                                                 
 hidden-to-output (Dense)    (None, 3, 3)              12        
                                                                 
Total params: 36
Trainable params: 36
Non-trainable params: 0
_________________________________________________________________


- Wx = 
- Wh = 3
- bias(hidden) = 
- Wy = 
- bias(output) = 

실제로 weight의 개수를 확인해봅시다. 

In [39]:
for model_weight in model.weights:
    print(model_weight.name, '=>', model_weight.shape)

SimpleRNN-1/simple_rnn_cell_10/kernel:0 => (4, 5)
SimpleRNN-1/simple_rnn_cell_10/recurrent_kernel:0 => (5, 5)
SimpleRNN-1/simple_rnn_cell_10/bias:0 => (5,)
hidden-to-output/kernel:0 => (5, 3)
hidden-to-output/bias:0 => (3,)


### 2. 다층 Simple RNN

임의의 시계열 데이터를 생성하겠습니다. 

In [40]:
import numpy as np

data = np.sin(2*np.pi*np.arange(0, 1000))*10 + np.random.random(1000)
data.shape

(1000,)

In [ ]:
plt.figure(figsize=[100, 20])
plt.plot(np.arange(0, 1000), data)
plt.show()

In [42]:
from sklearn.model_selection import train_test_split

X, y 데이터 생성하기 

다음으로, 해당 샘플 데이터를 1칸씩 이동하며 10개씩 데이터를 잘라, x값으로 사용하고, y값은 11번째 데이터로 사용하겠습니다.  
예) x = 1번째 ~ 10번째 데이터 / y = 11번째

질문) 길이가 1000인 data를 이용해, 몇 쌍의 x, y 데이터를 만들어낼 수 있을까요?


In [43]:
x_sin = np.array([[[data[i+j]] for j in range(10)] for i in range(len(data)-9-1)])
y_sin = np.array([[data[i+10]] for i in range(len(data)-10)])
x_sin.shape, y_sin.shape

((990, 10, 1), (990, 1))

In [ ]:
# x_train, x_test, y_train, y_test = train_test_split(x_sin, y_sin, test_size=0.25, random_state=42)

In [44]:
# shuffle = False로 명시하면, 데이터셋 순차적으로 split
x_train, x_test, y_train, y_test = train_test_split(x_sin, y_sin, shuffle=False) 

In [45]:
x_train.shape, x_test.shape, y_train.shape
# 742 : 데이터 개수, 10 : timestep, 1은 숫자 1개라는 의미

((742, 10, 1), (248, 10, 1), (742, 1))

### 모델 작성.
1. layer1: SimpleRNN 8
2. layer2: SimpleRNN 8
3. layer3: SimpleRNN 4
4. output_layer:  Dense 1 # 값이 하나 ManyToOne(회귀)

### RNN 층을 쌓을 때는 마지막 RNN 말고는 무조건 >> return_sequences = True
마지막 return_sequences = True는 문제에 따라 결정
many to many 는 True, many to one은 False

In [56]:
model = models.Sequential()

# 다층 LSTM 
model.add(layers.SimpleRNN(units = 8,
                           input_shape = (10,1),
                           return_sequences = True, # 다층이라서 True
                           name = 'rnn_1'))
model.add(layers.SimpleRNN(8,input_shape = (10,8),return_sequences = True, name = 'rnn_2')) 
# input_shape 적어줘야 한다
# 층을 쌓으려면 return_sequences는 True로 줘야 한다
model.add(layers.SimpleRNN(4,input_shape = (10,8),name = 'rnn_3')) # 숫자 맞추는 거라 return_sequences 안줌 

model.add(layers.Dense(1)) # activation=은 linear 혹은 안줘야 한다

model.compile(optimizer='adam', loss='mse')

model.fit(x_train, y_train, epochs=10, verbose=1)

Epoch 1/10
24/24 [==============================] - 3s 8ms/step - loss: 0.4359
Epoch 2/10
24/24 [==============================] - 0s 8ms/step - loss: 0.1629
Epoch 3/10
24/24 [==============================] - 0s 12ms/step - loss: 0.1069
Epoch 4/10
24/24 [==============================] - 0s 14ms/step - loss: 0.0964
Epoch 5/10
24/24 [==============================] - 0s 14ms/step - loss: 0.0924
Epoch 6/10
24/24 [==============================] - 0s 14ms/step - loss: 0.0902
Epoch 7/10
24/24 [==============================] - 0s 14ms/step - loss: 0.0887
Epoch 8/10
24/24 [==============================] - 0s 13ms/step - loss: 0.0876
Epoch 9/10
24/24 [==============================] - 0s 13ms/step - loss: 0.0869
Epoch 10/10
24/24 [==============================] - 0s 13ms/step - loss: 0.0864


In [ ]:
model.summary()

Model: "sequential_11"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 rnn-1 (SimpleRNN)           (None, 10, 8)             80        
                                                                 
 rnn-2 (SimpleRNN)           (None, 10, 8)             136       
                                                                 
 rnn-3 (SimpleRNN)           (None, 4)                 52        
                                                                 
 dense_3 (Dense)             (None, 1)                 5         
                                                                 
Total params: 273
Trainable params: 273
Non-trainable params: 0
_________________________________________________________________


- unit: 노드의 개수 (= 아웃풋 공간의 차원) 
- (layer 1) Wx + Wh + b = 8X1 = (8X1)X(1X1) + (8X8)X(8X1) + 8X1 = 8+64+8 >> 80

- (layer 2) Wx + Wh + b = 8X1 = (8X8)X(8X1) + (8X8)X(8X1) + 8X1 = 64+64+8 >> 136

- (layer 3) Wx + Wh + b = 4X1 = (4X8)X(8X1) + (4X4)X(4X1) + 4X1 = 32+16+4 >> 52


In [ ]:
for model_weight in model.weights:
    print(model_weight.name, '=>', model_weight.shape)

In [ ]:
model.weights

##### 학습결과

In [ ]:
y_hat = model.predict(x_test) #테스트값 예측

In [ ]:
y_axis = np.arange(0, len(y_train))
y_2_axis = np.arange(len(y_train), len(y_train)+len(y_hat))

plt.figure(figsize = (100,30))
# Train 데이터 그래프로 표현
plt.plot(y_axis, y_train.reshape(len(x_train),), '-o')
plt.plot(y_2_axis, y_hat.reshape(len(x_test),),'-o', color = 'red', label = 'predicted')
plt.plot(y_2_axis, y_test.reshape(len(x_test),), '-o', color = 'purple', alpha = 0.3, label = 'actual')
plt.legend(fontsize = 40)
plt.show()

### reference

- https://github.com/ukairia777/tensorflow-nlp-tutorial/blob/main/08.%20RNN/8-4.%20understanding_simplernn_and_lstm.ipynb